raster/src/main/scala/geotrellis/raster/summary/GridVisitor.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.raster.summary
+
+import geotrellis.raster.Grid
+
+/**
+  * Visitor used to summarize a two-dimensional grid T that stores values of R
+  *
+  * The user should implement concrete subclasses that update the value of `result` as
+  * necessary on each call to `visit(raster: T, col: Int, row: Int)`.
+  *
+  * Be sure to handle the empty state. This could occur if no points in T are ever visited.
+  *
+  * GridVisitor is contravariant in T and covariant in R in the same fashion as Function1.
+  * This allows more generic concrete Visitor class implementations
+  * to satisfy type constraints for T at the call site, and the inverse is true for R. As a
+  * contrived example using GeoTrellis types, where Raster is a subtype of CellGrid:
+  *
+  * val DoubleVisitor: GridVisitor[CellGrid[Int], Double] = ???
+  * def usesRaster(visitor: GridVisitor[Raster[Tile]], Any]): Unit = ???
+  * usesRaster(DoubleVisitor)
+  *
+  * will compile. For an alternate explanation on covariance and contravariance that might help
+  * to clarify this, see:
+  * - https://stackoverflow.com/a/10604305
+  * - https://stackoverflow.com/a/38577878
+  */
+trait GridVisitor[-T <: Grid[Int], +R] extends Serializable {
```

Comment:
Do you think its worth while bumping this up to `geotrellis.raster` ?

Reply:
So...I had it there. Then I moved it because I think I made the argument with myself that it's primary use is reducing values within the raster to single result -- a summary. 

I don't feel strongly about it either way though so if you think its more discoverable/reusable up a level in `geotrellis.raster` I'm happy to move it.

Reply:
I think your internal argument is good enough for me, lets leave it in here.

Reply:
👍